tests/unit/customizations/history/test_db.py

```diff
@@ -0,0 +1,711 @@
+
+        # Each thread should have its own set of request_ids so the request
+        # ids in each set of records should match, but should not match the
+        # request_ids from the other thread.
+        a_request_ids = [record[1] for record in a_records]
+        self.assertEqual(len(a_request_ids), 2)
+        self.assertEqual(a_request_ids[0], a_request_ids[1])
+        thread_a_request_id = a_request_ids[0]
+
+        b_request_ids = [record[1] for record in b_records]
+        self.assertEqual(len(b_request_ids), 2)
+        self.assertEqual(b_request_ids[0], b_request_ids[1])
+        thread_b_request_id = b_request_ids[0]
+
+        self.assertNotEqual(thread_a_request_id, thread_b_request_id)
+
+        # Since the request_id is reset by the API_CALL record being written
+        # and thread b has now written an API_CALL record the request id has
+        # been reset once. To ensure this doesnt bleed over to other threads
+        # we will write another record in thread a and ensure that it's
+        # request_id matches the previous ones from thread a rather than then
+        # thread b request_id
+        self.thread_a.write_record({
+            'event_type': 'HTTP_RESPONSE',
+            'payload': '',
+            'source': 'TEST_a',
+            'timestamp': 1234
+        })
+        self.thread_b.write_record({
+            'event_type': 'HTTP_RESPONSE',
+            'payload': '',
+            'source': 'TEST_b',
+            'timestamp': 1234
+        })
+        a_record = self.thread_a.read_n_records(1)[0]
+        b_record = self.thread_b.read_n_records(1)[0]
+        self.assertEqual(a_record[1], thread_a_request_id)
+        self.assertEqual(b_record[1], thread_b_request_id)
+
+
+class TestDatabaseRecordReader(BaseDatabaseRecordTester):
+    def setUp(self):
+        self.fake_connection = FakeDatabaseConnection()
+        self.reader = DatabaseRecordReader(self.fake_connection)
+
+    def test_row_factory_set(self):
+        self.assertEqual(self.fake_connection.row_factory,
+                         self.reader._row_factory)
+
+    def test_yield_latest_records(self):
+        pass
+
+    def test_yield_records(self):
+        pass
+
+
+class TestPayloadSerialzier(unittest.TestCase):
+    def test_can_serialize_basic_types(self):
+        original = {
+            'string': 'foo',
+            'int': 4,
+            'list': [1, 2, 'bar'],
+            'dict': {
+                'sun': 'moon'
+            },
+            'float': 1.2
+        }
+        string_value = json.dumps(original, cls=PayloadSerializer)
+        reloaded = json.loads(string_value)
+        self.assertEqual(original, reloaded)
+
+    def test_can_serialize_datetime(self):
+        now = datetime.datetime.now()
+        iso_now = now.isoformat()
+        string_value = json.dumps(now, cls=PayloadSerializer)
+        reloaded = json.loads(string_value)
+        self.assertEqual(iso_now, reloaded)
+
```

Comment:
Should we add a test for the unknown types?

Reply:
Yep.